Framework/PythonInterface/mantid/api/src/Exports/IFunction.cpp

```diff
@@ -35,6 +36,32 @@ PyObject *getCategories(IFunction &self) {
   return registered;
 }
 
+/// An object for constructing a shared_ptr that won't ever delete its pointee
+class NoDeleting {
+public:
+  /// Does nothing
+  void operator()(void *) {}
+  /// Does nothing
+  void operator()(const void *) {}
+};
+
+//------------------------------------------------------------------------------------------------------
+/**
+* Something that makes IFunction return to python a composite function
+* for Product function, Convolution or
+* any similar superclass of composite function.
+* @param self :: Enables it to be called as a member function on the
+* IFunction class
+*/
+Mantid::API::CompositeFunction_sptr castToCompositeFunction(IFunction &self) {
```

Comment:
Going back to the original purpose of this thread, @mantid-roman suggested the cast function, that we'd like to remove, to solve a serious problem with the ``*`` operator for composite functions.
We want
```
p = f1 * f2 *f3 
p = ProductFunctionWrapper( f1, f2, f3)
```
to have the same effect and in particular the component functions would be referred to by
`` p[0], p[1], p[2]``.
This does not occur from repeated use of the ``*`` operator and so the result needs every composite function within it replaced by its list of functions (which I refer to flattening).
When doing this within a ProductFunction, a composite function within it is presented as a IFunction. 

I think if we were to remove the cast and implement integer indexing and length as suggested, we'd at best get only the first function in the composite function and lose the others from the product.